go/api/config.go

```diff
@@ -0,0 +1,309 @@
+// Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+
+package api
+
+import "github.com/aws/glide-for-redis/go/glide/protobuf"
+
+const (
+	defaultHost = "localhost"
+	defaultPort = 6379
+)
+
+// NodeAddress represents the host address and port of a node in the cluster.
+type NodeAddress struct {
+	Host string // Optional: if not supplied, "localhost" will be used.
+	Port int    // Optional: if not supplied, 6379 will be used.
+}
+
+func (addr *NodeAddress) toProtobuf() *protobuf.NodeAddress {
+	if addr.Host == "" {
+		addr.Host = defaultHost
+	}
+
+	if addr.Port == 0 {
+		addr.Port = defaultPort
+	}
+
+	return &protobuf.NodeAddress{Host: addr.Host, Port: uint32(addr.Port)}
+}
+
+// RedisCredentials represents the credentials for connecting to a Redis server.
+type RedisCredentials struct {
+	// Optional: the username that will be used for authenticating connections to the Redis servers. If not supplied, "default"
+	// will be used.
+	username string
+	// Required: the password that will be used for authenticating connections to the Redis servers.
+	password string
+}
+
+// NewRedisCredentials returns a new RedisCredentials struct with the given username and password.
+func NewRedisCredentials(username string, password string) *RedisCredentials {
+	return &RedisCredentials{username, password}
+}
+
+// NewRedisCredentialsWithDefaultUsername returns a new RedisCredentials struct with a default username of "default" and the
+// given password.
+func NewRedisCredentialsWithDefaultUsername(password string) *RedisCredentials {
+	return &RedisCredentials{password: password}
+}
+
+func (creds *RedisCredentials) toProtobuf() *protobuf.AuthenticationInfo {
+	return &protobuf.AuthenticationInfo{Username: creds.username, Password: creds.password}
+}
+
+// ReadFrom represents the client's read from strategy.
+type ReadFrom int
+
+const (
+	// Primary - Always get from primary, in order to get the freshest data.
+	Primary ReadFrom = iota
+	// PreferReplica - Spread the requests between all replicas in a round-robin manner. If no replica is available, route the
+	// requests to the primary.
+	PreferReplica
+)
+
+func mapReadFrom(readFrom ReadFrom) protobuf.ReadFrom {
+	if readFrom == PreferReplica {
+		return protobuf.ReadFrom_PreferReplica
+	}
+
+	return protobuf.ReadFrom_Primary
+}
+
+type baseClientConfiguration struct {
+	addresses      []NodeAddress
```

Comment:
I've made these fields private (indicated in Go by the lower-case first letter) so that the user has to use the public With* builder methods instead of accessing them directly. Instead of putting docs for the private fields here, I've put the docs above the public With* builder methods below.